tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/ZipOracle.scala

```diff
@@ -0,0 +1,32 @@
+package at.forsyte.apalache.tla.bmcmt.rules.aux
+
+import at.forsyte.apalache.tla.bmcmt.SymbState
+import at.forsyte.apalache.tla.bmcmt.smt.SolverContext
+import at.forsyte.apalache.tla.lir.TypedPredefs.BuilderExAsTyped
+import at.forsyte.apalache.tla.lir.convenience.tla
+import at.forsyte.apalache.tla.lir.{BoolT1, TlaEx}
+
+/**
+ * {{{ZipOracle}}} is an optimization of {{{Oracle}}}. It groups several values of the background oracle together, in
+ * order to reduce the number of constraints. It is up to the user of {{{ZipOracle}}} to make sure that the grouped
+ * values may be treated as equivalent.
+ *
+ * @param backOracle
+ *   the background oracle whose values are grouped together
+ * @param groups
+ *   a list of groups over the values of the background oracle
+ */
+class ZipOracle(backOracle: Oracle, groups: List[Set[Int]]) extends Oracle {
```

Comment:
Shouldn't `ZipOracle` override `caseAssertions` is a similar fashion to `SparseOracle`?

Reply:
`ZipOracle` is using `whenEqualTo`, because the values should go over the groups, not the original values.

Reply:
Sorry to open this again, but is the assumption that `ZipOracle` will always be used properly? As is, `ZipOracle` will default to `Oracle`'s behavior for `caseAssertions`, which should not be the case from what I understand. Shouldn't `backOracle` be called?

Reply:
Right now it makes no difference, as we have the same body for all non-mock oracles, which you nicely refactored, but it may not always be the case. I'm trying to anticipate problems, but maybe I'm just overthinking it.

Reply:
I think problems may arise, if we mix oracles. If we always use `ZipOracle`, as `CherryPick` does, this should be fine.